Varnishadm: Add -e argument

walid-git opened the pull request:
> -e
    Exit immediately if a command fails in `pass` mode and return the CLI
    status code of the failing command. This has no effect on `interactive`
    mode (except when `-p` is used). Similarly to `CLI Command File` (see
    :ref:`varnishd(1)`), if a command is prefixed with '-', its failure will
    be ignored and will not cause varnishadm to exit.

Fixes: #4012

walid-git commented:
Bugwash: 

> [15:09:38] \<phk\> I dont see any problems
[15:09:49] \<phk\> but I guess we should give @nigoroll  a chance

Dridi commented on bin/varnishtest/tests/u00001.vtc:
```diff
@@ -61,3 +61,37 @@ shell -err {
+	EOF
+}
+
+shell { varnishadm -n ${v1_name} < cli_file.txt }
+shell { test 2 -eq $(varnishadm -n ${v1_name} < cli_file.txt | grep "Child in state" | wc -l) }
+
+shell -exit 101 { varnishadm -e -n ${v1_name} < cli_file.txt }
+shell { test 1 -eq $(varnishadm -e -n ${v1_name} < cli_file.txt | grep "Child in state" | wc -l) }
+
+shell { cat cli_file.txt | varnishadm -n ${v1_name} }
+shell { test 2 -eq $(cat cli_file.txt | varnishadm -n ${v1_name} | grep "Child in state" | wc -l) }
+
+shell -exit 101 { cat cli_file.txt | varnishadm -e -n ${v1_name} }
+shell { test 1 -eq $(cat cli_file.txt | varnishadm -e -n ${v1_name} | grep "Child in state" | wc -l) }
```
I think we can do something simpler:
- run `varnishadm <file` without and with `-e`
  - no need to pipe with `cat` 
- don't use `status` twice, for example `ping`
- inspect the output accordingly

walid-git replied:
I tested both `varnishadm.. < file` and the `cat file | varnishadm..` versions on purpose, that is because, for some reason, the bug fixed in 9dfc726172b2f411bdd5fbae6d8187cd216f2384 only affected the `cat file | varnishadm..` version and not the other one, so I thought it wouldn't hurt having some coverage for both.

Dridi replied:
I see, regular file vs pipe... Then let's keep both, but the other suggestions apply.

walid-git replied:
Fixed: Replaced the second status with vcl.list (which is easier to check compared to ping response that gives different output on each call), and added output inspection with diff

walid-git commented on lib/libvarnish/vcli_serve.c:
```diff
@@ -599,20 +599,18 @@ VCLS_Poll(struct VCLS *cs, const struct cli *cli, int timeout)
+			h = i;
 	}
-	return (k);
+	return (h ? h : k);
```
Or, could we simply ignore `POLLHUP` and only rely on read returning 0 ?

Dridi replied:
Or rather return `<= 0` :)

walid-git replied:
Removed the `POLLHUP` check. Now we only rely on the `read` return value.

walid-git commented on bin/varnishadm/varnishadm.c:
```diff
@@ -341,6 +342,15 @@ interactive(int sock)
+		    cli->result, VSB_data(cli->sb));
+		RL_EXIT(cli->result);
+	}
+}
```
Oh, sorry about the code style, i'll fix that.  
Regarding the name, we have two callbacks for each VCLS server, one is called before executing the command and the other is called after, and we use this naming ( `*_cb_before` and `*_cb_after`) in both mgt and cache cli.

Dridi replied:
Then keep the name as it is.

walid-git replied:
fixed

walid-git commented on bin/varnishadm/varnishadm.c:
```diff
@@ -512,6 +523,9 @@ main(int argc, char * const *argv)
 			break;
+		case 'e':
+			e_arg = 1;
+			break;
```
fixed

walid-git commented on doc/sphinx/reference/varnishadm.rst:
```diff
@@ -62,6 +62,12 @@ OPTIONS
 -t timeout
     Wait no longer than this many seconds for an operation to finish.
 
+-e
```
fixed

walid-git commented on doc/sphinx/reference/varnishadm.rst:
```diff
@@ -19,7 +19,7 @@ Control a running Varnish instance
 ========
 
-varnishadm [-h] [-n ident] [-p] [-S secretfile] [-T [address]:port] [-t timeout] [command [...]]
+varnishadm [-h] [-n ident] [-p] [-S secretfile] [-T [address]:port] [-t timeout] [-e] [command [...]]
```
fixed

walid-git commented on bin/varnishadm/varnishadm.c:
```diff
@@ -406,7 +417,7 @@ usage(int status)
 	fprintf(stderr,
 	    "Usage: varnishadm [-h] [-n ident] [-p] [-S secretfile] "
-	    "[-T [address]:port] [-t timeout] [command [...]]\n");
+	    "[-T [address]:port] [-t timeout] [-e] [command [...]]\n");
```
fixed

walid-git commented on bin/varnishadm/varnishadm.c:
```diff
@@ -512,6 +523,9 @@ main(int argc, char * const *argv)
 			break;
+		case 'e':
+			e_arg = 1;
+			break;
```
fixed

Dridi commented on bin/varnishadm/varnishadm.c:
```diff
@@ -338,51 +340,66 @@ interactive(int sock)
 	}
 }
 
-/*
- * No arguments given, simply pass bytes on stdin/stdout and CLI socket
- */
```
Please retain this comment for now to improve the diff.

Dridi commented on bin/varnishadm/varnishadm.c:
```diff
@@ -338,51 +340,66 @@ interactive(int sock)
+	for (i = 1; av[i] != NULL; i++) {
+		VSB_quote(cli_buf, av[i], strlen(av[i]), 0);
+		VSB_putc(cli_buf, ' ');
+	}
```
I think now is a good time to tackle the lack of SP quoting and share the logic with the `do_args()` function.

This could mean a new `VSB_QUOTE_CLI` flag. I'll let you think about our options here. One could be to always write double-quoted arguments with such a flag, except of course when we have a here-document.

Dridi reviewed (commented):
Since we are in the neighborhood...

Dridi commented:
> If it is not too much work, could it be split into smaller commits or could the commit message explain better what it does?

Not sure how the github web UI would render it, but if the discarded comment was preserved as I suggested we would likely have a cleaner diff.

walid-git commented:
I brought back the comment mentioned by Dridi, hopefully this will make the diff a bit more readable. Otherwise, I don't see any way to break the commit into smaller ones.

walid-git commented:
- Changed the return value of `varnishadm -e` to  `cli status / 100` after phk's suggestion during bugwash. That is because CLI status can go beyond 255 and we might have a status code that is N*256 that will return 0.
- Added polling to the remote cli socket so that we are notified if the socket hangs up when we are not writing anything.
- Added quoting to the CLI command arguments so that the VCLS server we are forwarding commands to sees the arguments grouped in the same way they were received.

Note that test will not pass until the bugfix in f354ca99f133fb501d01849ca8a111981c15fb84 is merged.

nigoroll commented on lib/libvarnish/vsb.c:
```diff
@@ -529,7 +529,7 @@ VSB_quote_pfx(struct vsb *s, const char *pfx, const void *v, int len, int how)
 	const uint8_t *p = v;
 	const uint8_t *q;
-	int quote = 0;
+	int quote = 0, dq = 0;
```
Flexelint:

> Info 838: Previously assigned value to variable 'dq' has not been used

Just remove the initialization please.

walid-git replied:
fixed

Dridi commented on include/vsb.h:
```diff
@@ -123,6 +123,11 @@ void		 VSB_destroy(struct vsb **);
 	 * Not valid with VSB_QUOTE_JSON and VSB_QUOTE_HEX
 	 */
 
+#define VSB_QUOTE_CLI		64
+	/*
+	 * Add quotes ".." around arguments when needed
+	 */
```
I think the rule for "when needed" is the presence of either space, tabulation, or double quote in the argument value, so in its current state `VSB_QUOTE_CLI` is insufficient.

walid-git commented on bin/varnishadm/varnishadm.c:
```diff
@@ -433,7 +437,13 @@ pass(int sock)
+		if (j == 1 && recv(sock, buf, 1, MSG_PEEK) == 0) {
+			fprintf(stderr, "CLI communication error\n");
+			RL_EXIT(1);
+		}
```
I used peek, because, on my system at least, poll would report POLLIN event when the remote socket is closed, and that could not be differentiated from when we have actual data to read (which should never be the case but still..), so I found using recv with MSG_PEEK to be the most reliable solution here as it will always return 0 on EOF

walid-git commented:
Addressed dridi's comments:

- Harmonized the exit status and documented them in EXIT STATUS section of the man page.
- We don't forward the parsed heredoc now, we reconstruct it and forward it to be parsed in mgt instead.

Ready for review

Dridi reviewed (changes requested):
We looked at some of the issues brought up in this pull request with @walid-git and we are considering adding small building blocks to `vcli_serve` to offer a "CLI proxy" feature. @walid-git is working on it, and meanwhile I forgot to submit the small nitpicks we found while reviewing the commits.
